### PR TITLE
If you have a stronger acid, you can now click on weaker acid directly to upgrade it.

### PR DESCRIPTION
**Pull request by ZeWaka**
## About The Pull Request

If you have a stronger acid, you can now click on weaker acid directly to upgrade it.
Fixes a bug with acid upgrading not transferring the existing ticks

## Why It's Good For The Game

Huge acid QoL, especially for shit like flares and their animated sprite
Bug fix good

image

## Changelog
:cl:
add: If you have a stronger acid, you can now click on weaker acid directly to upgrade it.
fix: Fixed acid upgrading not inheriting the existing melt progress
/:cl:

### CODE REVIEW

File: code/game/objects/effects/aliens.dm [hunk 1]
```diff
@@ -149,10 +149,11 @@
 	acid_damage = 175
 	icon_state = "acid_strong"
 
-/obj/effect/xenomorph/acid/Initialize(mapload, target, melting_rate)
+/obj/effect/xenomorph/acid/Initialize(mapload, target, melting_rate, existing_ticks)
 	. = ..()
 	acid_melt_multiplier = melting_rate
 	acid_t = target
+	ticks += existing_ticks
```
Comment (ZeWaka) [inline code/game/objects/effects/aliens.dm:156]: `+=` because `0 += null == 0`

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 2]
```diff
@@ -596,6 +596,10 @@
 	use_state_flags = XACT_USE_BUCKLED
 
 /datum/action/xeno_action/activable/corrosive_acid/can_use_ability(atom/A, silent = FALSE, override_flags)
+	// Check if it's an acid object we're upgrading
+	if (istype(A, /obj/effect/xenomorph/acid))
+		var/obj/effect/xenomorph/acid/existing_acid = A
+		A = existing_acid.acid_t // Swap the target to the target of the acid
```
Comment (ZeWaka) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:602]: Yes, the duplicate check is needed, as this is checked throughout.

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 3]
```diff
@@ -641,26 +652,6 @@
 	span_xenowarning("We vomit globs of vile stuff all over \the [A]. It begins to sizzle and melt under the bubbling mess of acid!"), null, 5)
 	playsound(X.loc, "sound/bullets/acid_impact1.ogg", 25)
 
-/datum/action/xeno_action/activable/corrosive_acid/proc/acid_progress_transfer(acid_type, obj/O, turf/T)
```
Comment (ZeWaka) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:644]: Unused, from ancient days - see the turf check.

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 4]
```diff
@@ -630,8 +640,9 @@
 	if(!can_use_ability(A, TRUE))
 		return fail_activate()
 
+	var/old_acid_ticks = A.current_acid?.ticks
```
Comment (ZeWaka) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:643]: Done this way because it'll be null if the object isn't already an acid target

**Comment by novaepee** [comment]
good meme

**Comment by Strayed-Raven** [comment]
Is this any xenos with stronger acid being able to upgrade a weaker stack. Or is this giving weak acid xenos the ability to self upgrade after another channel?

**Comment by ZeWaka** [comment]
> Is this any xenos with stronger acid being able to upgrade a weaker stack.

This. They already can by clicking the actual object being acided, but this lets them click the acid object which is generally much bigger and on top.

**Pull request merged**